R/client.R

```diff
@@ -0,0 +1,413 @@
+# This file implements R as the _client_.
+
+# tools ------------------------------------------------------------------------
+# each named entry is:
+# name: the name of the server (from the config)
+# process: the processx process object
+# tools: a named list of tools
+# id: the current jsonrpc message id
+the$mcp_servers <- list()
+
+#' Define ellmer tools from MCP servers
+#'
+#' @description
+#' These functions implement R as an MCP _client_, so that ellmer chats can
+#' register functionality from third-party MCP servers such as those listed
+#' here: <https://github.com/modelcontextprotocol/servers>.
+#'
+#' `mcp_tools()` fetches tools from MCP servers configured in the acquaint
+#' server config file at `set_mcp_config()` and converts them to a list of
+#' tools compatible with the `$set_tools()` method of [ellmer::Chat] objects.
+#'
+#' @section Configuration:
+#'
+#' acquaint uses the same .json configuration file format as Claude Desktop;
+#' most MCP servers will define example .json to configure the server with
+#' Claude Desktop in their README files. By default, acquaint will look to
+#' `file.path("~", ".config", "acquaint", "config.json")`; you can edit that
+#' file with `file.edit(file.path("~", ".config", "acquaint", "config.json"))`,
+#' or you can supply an alternative file path with `set_mcp_config()`.
+#'
+#' The acquaint config file should be valid .json with an entry `mcpServers`.
+#' That entry should contain named elements, each with at least a `command`
+#' and `args` entry.
+#'
+#' For example, to configure `mcp_tools()` with GitHub's official MCP Server
+#' <https://github.com/github/github-mcp-server>, you could write the following
+#' in that file:
+#'
+#' ```json
+#' {
+#'   "mcpServers": {
+#'     "github": {
+#'       "command": "docker",
+#'       "args": [
+#'         "run",
+#'         "-i",
+#'         "--rm",
+#'         "-e",
+#'         "GITHUB_PERSONAL_ACCESS_TOKEN",
+#'         "ghcr.io/github/github-mcp-server"
+#'       ],
+#'       "env": {
+#'         "GITHUB_PERSONAL_ACCESS_TOKEN": "<add_your_github_pat_here>"
+#'       }
+#'     }
+#'   }
+#' }
+#' ```
+#'
+#' @returns
+#' * `mcp_tools()` returns a list of ellmer tools that can be passed directly
+#' to the `$set_tools()` method of an [ellmer::Chat] object.
+#' * `set_mcp_config()` returns the path to the new MCP config file, invisibly.
+#'
+#' @name client
+#' @export
+mcp_tools <- function() {
+  config <- read_mcp_config()
```

Comment:
I'm not a huge fan of the pattern of `set/get/read_mcp_config()` – although I may be missing some context. I'd rather expose a `path_config` argument in `mcp_client_tools()` that can be passed through to wherever it's used. 

If the user needs to be able to set a global option, I'd recommend using an R option, documented in `mcp_client_config()` or somewhere similarly relevant.

Personally, I'd use `mcp_client_config(path = NULL)` to read the client config at `path`, the global option, or the default path. In that world, you could have `mcp_client_tools(config = mcp_client_config())` by default, but users could provide `config` with `mcp_client_config("my-config.json")`.